SwiftLSPClient/Types/Basic.swift

```diff
@@ -83,6 +83,20 @@ public struct TextDocumentItem: Codable {
         self.version = version
         self.text = text
     }
+
+    public init(path: String) throws {
+        let url = URL(fileURLWithPath: path)
+        self.uri = url.absoluteString
+        let ext = url.pathExtension
+        guard let languageID = LanguageIdentifier
+            .fileExtensions[ext] else {
+                throw NSError(domain: "SwiftLSPClient", code: -1,
+                              userInfo: ["Invalid source type \(ext)": path])
+        }
+        self.languageId = languageID
+        self.version = 1
+        self.text = try String(contentsOf: url)
+    }
```

Comment:
There's a lot happening in this method. How about something like this:

- add an init similar to what do you with `TextDocumentIdentifier`
- add a static method to `LanguageIdentifier` to compute a languageID given a url
- finally, wrap this all up in something like `init(contentsAt: path String, version: Int = 1)`
- Also might be nicer to wrap up that error into an Error type. A little more swifty, and makes it easier to surface the contextual information (that the extension isn't in that dictionary)

This should achieve exactly what you are after, keep some consistency with the change to `TextDocumentIdentifier` and also add some flexibility for other uses.

Reply:
I don't know if you are still awake but I've pushed another commit looking to address what you are after and the code seems to be the better for it.

Reply:
I just saw another thread crash (1 in a day) so my locks aren't quite up to scratch.